File: mobile/src/main/java/rs/readahead/washington/mobile/views/activity/AudioPlayActivity.java
//package rs.readahead.washington.mobile.views.activity;
//
//import static com.hzontal.tella_vault.Metadata.VIEW_METADATA;
//import static rs.readahead.washington.mobile.views.fragment.vault.attachements.AttachmentsFragmentKt.PICKER_FILE_REQUEST_CODE;
//
//import android.Manifest;
//import android.app.ProgressDialog;
//import android.content.Context;
//import android.content.Intent;
//import android.os.Build;
//import android.os.Bundle;
//import android.os.Handler;
//import android.view.Menu;
//import android.view.MenuItem;
//import android.view.View;
//import android.view.WindowManager;
//import android.widget.ImageButton;
//import android.widget.TextView;
//
//import androidx.annotation.Nullable;
//import androidx.appcompat.app.AlertDialog;
//import androidx.appcompat.widget.Toolbar;
//
//import com.hzontal.tella_vault.VaultFile;
//
//import org.hzontal.shared_ui.bottomsheet.BottomSheetUtils;
//import org.hzontal.shared_ui.bottomsheet.VaultSheetUtils;
//
//import java.util.LinkedHashMap;
//import java.util.Locale;
//import java.util.concurrent.TimeUnit;
//
//import kotlin.Unit;
//import permissions.dispatcher.NeedsPermission;
//import permissions.dispatcher.OnNeverAskAgain;
//import permissions.dispatcher.OnPermissionDenied;
//import permissions.dispatcher.OnShowRationale;
//import permissions.dispatcher.PermissionRequest;
//import permissions.dispatcher.RuntimePermissions;
//import rs.readahead.washington.mobile.MyApplication;
//import rs.readahead.washington.mobile.R;
//import rs.readahead.washington.mobile.bus.event.MediaFileDeletedEvent;
//import rs.readahead.washington.mobile.bus.event.VaultFileRenameEvent;
//import rs.readahead.washington.mobile.databinding.ActivityAudioPlayBinding;
//import rs.readahead.washington.mobile.media.AudioPlayer;
//import rs.readahead.washington.mobile.media.MediaFileHandler;
//import rs.readahead.washington.mobile.mvp.contract.IAudioPlayPresenterContract;
//import rs.readahead.washington.mobile.mvp.contract.IMediaFileViewerPresenterContract;
//import rs.readahead.washington.mobile.mvp.presenter.AudioPlayPresenter;
//import rs.readahead.washington.mobile.mvp.presenter.MediaFileViewerPresenter;
//import rs.readahead.washington.mobile.util.DialogsUtil;
//import rs.readahead.washington.mobile.util.PermissionUtil;
//import rs.readahead.washington.mobile.util.ThreadUtil;
//import rs.readahead.washington.mobile.views.base_ui.BaseLockActivity;
//import rs.readahead.washington.mobile.views.fragment.vault.info.VaultInfoFragment;
//import timber.log.Timber;
//
//@RuntimePermissions
//public class AudioPlayActivity extends BaseLockActivity implements
//        IAudioPlayPresenterContract.IView,
//        IMediaFileViewerPresenterContract.IView {
//    public static final String PLAY_MEDIA_FILE = "pmf";
//    public static final String PLAY_MEDIA_FILE_ID_KEY = "pmfik";
//    public static final String NO_ACTIONS = "na";
//    private static final String TIME_FORMAT = "%02d:%02d:%02d";
//    private static final int SEEK_DELAY = 15000;
//
//    ImageButton mPlay;
//    ImageButton mRwd;
//    ImageButton mFwd;
//    TextView mTimer;
//    TextView mDuration;
//    View forward;
//    View rewind;
//
//    private AudioPlayPresenter presenter;
//    private VaultFile handlingVaultFile;
//    private AudioPlayer audioPlayer;
//    private AudioPlayer.Listener audioPlayerListener;
//
//    private MediaFileViewerPresenter viewerPresenter;
//
//    private boolean showActions = false;
//    private boolean actionsDisabled = false;
//    private boolean withMetadata = false;
//    private AlertDialog alertDialog;
//    private ProgressDialog progressDialog;
//
//    private boolean paused = true;
//    private Toolbar toolbar;
//    private boolean isInfoShown = false;
//    private ActivityAudioPlayBinding binding;
//
//
//    @Override
//    protected void onCreate(Bundle savedInstanceState) {
//        super.onCreate(savedInstanceState);
//
//        binding = ActivityAudioPlayBinding.inflate(getLayoutInflater());
//        setContentView(binding.getRoot());
//        initView();
//        initListeners();
//
//        overridePendingTransition(R.anim.slide_in_start, R.anim.fade_out);
//
//        setSupportActionBar(toolbar);
//        viewerPresenter = new MediaFileViewerPresenter(this);
//        enablePlay();
//
//        if (getIntent().hasExtra(NO_ACTIONS)) {
//            actionsDisabled = true;
//        }
//
//        audioPlayerListener = new AudioPlayer.Listener() {
//
//            @Override
//            public void onStart(int duration) {
//                mDuration.setText(timeToString(duration));
//            }
//
//            @Override
//            public void onStop() {
//                stopPlayer();
//                paused = true;
//                enablePlay();
//                showTimeRemaining(0);
//            }
//
//            @Override
//            public void onProgress(int currentPosition) {
//                showTimeRemaining(currentPosition);
//            }
//
//            private void showTimeRemaining(int left) {
//                mTimer.setText(timeToString(left));
//            }
//        };
//
//        if (getIntent().hasExtra(PLAY_MEDIA_FILE)) {
//            VaultFile vaultFile = (VaultFile) getIntent().getSerializableExtra(PLAY_MEDIA_FILE);
//            if (vaultFile != null) {
//                ThreadUtil.runOnMain(() -> onMediaFileSuccess(vaultFile));
//            }
//        } else if (getIntent().hasExtra(PLAY_MEDIA_FILE_ID_KEY)) {
//            String id = getIntent().getStringExtra(PLAY_MEDIA_FILE_ID_KEY);
//            if (id != null) {
//                presenter = new AudioPlayPresenter(this);
//                presenter.getMediaFile(id);
//            }
//        }
//    }
//
//    @Override
//    public void finish() {
//        super.finish();
//        overridePendingTransition(R.anim.slide_in_end, R.anim.slide_out_start);
//    }
//
//    @Override
//    public boolean onCreateOptionsMenu(Menu menu) {
//        if (!actionsDisabled && showActions) {
//            toolbar.inflateMenu(R.menu.video_view_menu);
//
//            if (handlingVaultFile != null && handlingVaultFile.metadata != null) {
//                MenuItem item = toolbar.getMenu().findItem(R.id.menu_item_metadata);
//                item.setVisible(true);
//            }
//        }
//
//        return super.onCreateOptionsMenu(menu);
//    }
//
//    @Override
//    public boolean onOptionsItemSelected(MenuItem item) {
//        int id = item.getItemId();
//
//        if (id == android.R.id.home) {
//            onBackPressed();
//            return true;
//        }
//
//        if (id == R.id.menu_item_more) {
//            showVaultActionsDialog(handlingVaultFile);
//            return true;
//        }
//
//        if (id == R.id.menu_item_metadata) {
//            showMetadata();
//            return true;
//        }
//
//        return super.onOptionsItemSelected(item);
//    }
//
//    private void initListeners() {
//        mPlay.setOnClickListener((view) -> {
//            if (paused) {
//                handlePlay();
//            } else {
//                handlePause();
//            }
//        });
//
//        forward.setOnClickListener((view) -> audioPlayer.ffwd(SEEK_DELAY));
//
//        rewind.setOnClickListener((view) -> audioPlayer.rwd(SEEK_DELAY));
//    }
//
//    @Override
//    public void onBackPressed() {
//        super.onBackPressed();
//        if (isInfoShown) {
//            toolbar.getMenu().findItem(R.id.menu_item_more).setVisible(true);
//            toolbar.getMenu().findItem(R.id.menu_item_metadata).setVisible(true);
//            toolbar.setTitle(handlingVaultFile.name);
//        } else {
//            stopPlayer();
//            finish();
//        }
//    }
//
//    @Override
//    protected void onResume() {
//        super.onResume();
//    }
//
//    @Override
//    protected void onPause() {
//        super.onPause();
//        handlePause();
//    }
//
//    @Override
//    protected void onDestroy() {
//        audioPlayerListener = null;
//
//        if (alertDialog != null && alertDialog.isShowing()) {
//            alertDialog.dismiss();
//        }
//
//        hideProgressDialog();
//
//        if (presenter != null) {
//            presenter.destroy();
//        }
//
//        if (viewerPresenter != null) {
//            viewerPresenter.destroy();
//            viewerPresenter = null;
//        }
//
//        super.onDestroy();
//    }
//
//
//    /*@SuppressLint("NeedOnRequestPermissionsResult")
//    @Override
//    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
//        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
//        AudioPlayActivityPermissionsDispatcher.onRequestPermissionsResult(this, requestCode, grantResults);
//    }*/
//    /*to check */
//    @OnPermissionDenied(Manifest.permission.WRITE_EXTERNAL_STORAGE)
//    void onWriteExternalStoragePermissionDenied() {
//    }
//    /*to check */
//    @OnNeverAskAgain(Manifest.permission.WRITE_EXTERNAL_STORAGE)
//    void onWriteExternalStorageNeverAskAgain() {
//    }
//    /*to check */
//    @NeedsPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)
//    void exportMediaFile() {
//        if (handlingVaultFile != null && viewerPresenter != null) {
//            if (handlingVaultFile.metadata != null) {
//                showExportWithMetadataDialog();
//            } else {
//                withMetadata = false;
//                maybeChangeTemporaryTimeout(() -> {
//                    performFileSearch();
//                    return Unit.INSTANCE;
//                });
//            }
//        }
//    }
//
//    private void performFileSearch() {
//        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.LOLLIPOP) {
//            Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT_TREE);
//            startActivityForResult(intent, PICKER_FILE_REQUEST_CODE);
//        } else {
//            viewerPresenter.exportNewMediaFile(withMetadata, handlingVaultFile, null);
//        }
//    }
///*to check */
//    @OnShowRationale(Manifest.permission.WRITE_EXTERNAL_STORAGE)
//    void showWriteExternalStorageRationale(final PermissionRequest request) {
//        maybeChangeTemporaryTimeout();
//        alertDialog = PermissionUtil.showRationale(this, request, getString(R.string.permission_dialog_expl_device_storage));
//    }
//
//    @Override
//    public void onMediaFileSuccess(VaultFile vaultFile) {
//        handlingVaultFile = vaultFile;
//        toolbar.setTitle(vaultFile.name);
//        //handlePlay();
//
//        if (!actionsDisabled) {
//            showActions = true;
//            invalidateOptionsMenu();
//        }
//    }
//
//    @Override
//    public void onMediaFileError(Throwable error) {
//        Timber.d(error, getClass().getName());
//    }
//
//    @Override
//    public void onMediaExported() {
//        showToast(getResources().getQuantityString((R.plurals.gallery_toast_files_exported), 1, 1));
//    }
//
//    @Override
//    public void onExportError(Throwable error) {
//        showToast(R.string.gallery_toast_fail_exporting_to_device);
//    }
//
//    @Override
//    public void onExportStarted() {
//        progressDialog = DialogsUtil.showProgressDialog(this, getString(R.string.gallery_save_to_device_dialog_progress_expl));
//    }
//
//    @Override
//    public void onExportEnded() {
//        hideProgressDialog();
//    }
//
//    @Override
//    public void onMediaFileDeleteConfirmation(VaultFile vaultFile, Boolean showConfirmDelete) {
//        if (showConfirmDelete) {
//            BottomSheetUtils.showConfirmSheet(
//                    getSupportFragmentManager(),
//                    getString(R.string.Vault_Warning_Title),
//                    getString(R.string.Vault_Confirm_delete_Description),
//                    getString(R.string.Vault_Delete_anyway),
//                    getString(R.string.action_cancel),
//                    isConfirmed -> {
//                        if (isConfirmed) {
//                            viewerPresenter.deleteMediaFiles(vaultFile);
//                        }
//                    }
//            );
//        } else {
//            viewerPresenter.deleteMediaFiles(vaultFile);
//        }
//    }
//
//    @Override
//    public void onMediaFileDeleted() {
//        MyApplication.bus().post(new MediaFileDeletedEvent());
//        finish();
//    }
//
//    @Override
//    public void onMediaFileDeletionError(Throwable throwable) {
//        showToast(R.string.gallery_toast_fail_deleting_files);
//    }
//
//    @Override
//    public void onMediaFileRename(VaultFile vaultFile) {
//        toolbar.setTitle(vaultFile.name);
//        MyApplication.bus().post(new VaultFileRenameEvent());
//    }
//
//    @Override
//    public void onMediaFileRenameError(Throwable throwable) {
//
//    }
//
//    @Override
//    public Context getContext() {
//        return this;
//    }
//
//    private void shareMediaFile() {
//        if (handlingVaultFile == null) {
//            return;
//        }
//
//        if (handlingVaultFile.metadata != null) {
//            showShareWithMetadataDialog();
//        } else {
//            startShareActivity(false);
//        }
//    }
//
//    private void startShareActivity(boolean includeMetadata) {
//        if (handlingVaultFile == null) {
//            return;
//        }
//
//        MediaFileHandler.startShareActivity(this, handlingVaultFile, includeMetadata);
//    }
//
//    private void showExportDialog() {
//      /*  alertDialog = DialogsUtil.showExportMediaDialog(this, (dialog, which) ->
//                AudioPlayActivityPermissionsDispatcher.exportMediaFileWithPermissionCheck(AudioPlayActivity.this));*/
//    }
//
//    private void showDeleteMediaDialog() {
//        alertDialog = new AlertDialog.Builder(this)
//                .setTitle(R.string.gallery_delete_files_dialog_title)
//                .setMessage(R.string.gallery_delete_files_dialog_expl)
//                .setPositiveButton(R.string.action_delete, (dialog, which) -> {
//                    if (viewerPresenter != null && handlingVaultFile != null) {
//                        viewerPresenter.deleteMediaFiles(handlingVaultFile);
//                    }
//                })
//                .setNegativeButton(R.string.action_cancel, (dialog, which) -> {
//                })
//                .setCancelable(true)
//                .show();
//    }
//
//    /*private void handleStop() {
//        stopPlayer();xr
//    }*/
//
//    private void handlePlay() {
//        if (handlingVaultFile == null) {
//            return;
//        }
//
//        if (audioPlayer != null) {
//            audioPlayer.resume();
//        } else {
//            audioPlayer = new AudioPlayer(this, audioPlayerListener);
//            audioPlayer.play(handlingVaultFile);
//        }
//
//        paused = false;
//        disablePlay();
//        disableScreenTimeout();
//    }
//
//    private void handlePause() {
//        if (handlingVaultFile == null) {
//            return;
//        }
//
//        enablePlay();
//        paused = true;
//
//        if (audioPlayer != null) {
//            audioPlayer.pause();
//        }
//        enableScreenTimeout();
//    }
//
//    private void onPlayerStop() {
//        enablePlay();
//    }
//
//    private void disablePlay() {
//        mPlay.setImageDrawable(getContext().getResources().getDrawable(R.drawable.big_white_pause_24p));
//        enableButton(forward, mFwd);
//        enableButton(rewind, mRwd);
//    }
//
//    private void enablePlay() {
//        mPlay.setImageDrawable(getContext().getResources().getDrawable(R.drawable.ic_play_arrow_white_24dp));
//        disableButton(forward, mFwd);
//        disableButton(rewind, mRwd);
//    }
//
//    private void enableButton(View view, ImageButton button) {
//        button.setClickable(true);
//        view.setAlpha(1f);
//    }
//
//    private void disableButton(View view, ImageButton button) {
//        button.setClickable(false);
//        view.setAlpha(.3f);
//    }
//
//    private void stopPlayer() {
//        if (audioPlayer != null) {
//            audioPlayer.stop();
//            audioPlayer = null;
//            onPlayerStop();
//            enableScreenTimeout();
//        }
//    }
//
//    private void hideProgressDialog() {
//        if (progressDialog != null) {
//            progressDialog.dismiss();
//            progressDialog = null;
//        }
//    }
//
//    private void showMetadata() {
//        Intent viewMetadata = new Intent(this, MetadataViewerActivity.class);
//        viewMetadata.putExtra(VIEW_METADATA, handlingVaultFile);
//        startActivity(viewMetadata);
//    }
//
//    private String timeToString(long duration) {
//        return String.format(Locale.ROOT, TIME_FORMAT,
//                TimeUnit.MILLISECONDS.toHours(duration),
//                TimeUnit.MILLISECONDS.toMinutes(duration) -
//                        TimeUnit.MINUTES.toMinutes(TimeUnit.MILLISECONDS.toHours(duration)),
//                TimeUnit.MILLISECONDS.toSeconds(duration) -
//                        TimeUnit.MINUTES.toSeconds(TimeUnit.MILLISECONDS.toMinutes(duration)));
//    }
//
//    private void disableScreenTimeout() {
//        getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
//    }
//
//    private void enableScreenTimeout() {
//        getWindow().clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
//    }
//
//    private void showVaultActionsDialog(VaultFile vaultFile) {
//        VaultSheetUtils.showVaultActionsSheet(getSupportFragmentManager(),
//                vaultFile.name,
//                getString(R.string.Vault_Upload_SheetAction),
//                getString(R.string.Vault_Share_SheetAction),
//                getString(R.string.Vault_Move_SheetDesc),
//                getString(R.string.Vault_Rename_SheetAction),
//                getString(R.string.gallery_action_desc_save_to_device),
//                getString(R.string.Vault_File_SheetAction),
//                getString(R.string.Vault_Delete_SheetAction),
//                false,
//                false,
//                false,
//                false,
//                new VaultSheetUtils.IVaultActions() {
//                    @Override
//                    public void upload() {
//
//                    }
//
//                    @Override
//                    public void share() {
//                        maybeChangeTemporaryTimeout(() -> {
//                            shareMediaFile();
//                            return Unit.INSTANCE;
//                        });
//                    }
//
//                    @Override
//                    public void move() {
//
//                    }
//
//                    @Override
//                    public void rename() {
//                        VaultSheetUtils.showVaultRenameSheet(
//                                getSupportFragmentManager(),
//                                getString(R.string.Vault_RenameFile_SheetTitle),
//                                getString(R.string.action_cancel),
//                                getString(R.string.action_ok),
//                                AudioPlayActivity.this,
//                                vaultFile.name,
//                                (name) -> {
//                                    viewerPresenter.renameVaultFile(vaultFile.id, name);
//                                    return Unit.INSTANCE;
//                                }
//                        );
//                    }
//
//                    @Override
//                    public void save() {
//                        BottomSheetUtils.showConfirmSheet(
//                                getSupportFragmentManager(),
//                                getString(R.string.gallery_save_to_device_dialog_title),
//                                getString(R.string.gallery_save_to_device_dialog_expl),
//                                getString(R.string.action_save),
//                                getString(R.string.action_cancel),
//                                isConfirmed -> {/*AudioPlayActivityPermissionsDispatcher.exportMediaFileWithPermissionCheck(AudioPlayActivity.this)*/}
//                        );
//                    }
//
//                    @Override
//                    public void info() {
//                        toolbar.setTitle(getString(R.string.Vault_FileInfo));
//                        toolbar.getMenu().findItem(R.id.menu_item_more).setVisible(false);
//                        toolbar.getMenu().findItem(R.id.menu_item_metadata).setVisible(false);
//                        invalidateOptionsMenu();
//                        addFragment(new VaultInfoFragment().newInstance(vaultFile, false), R.id.root);
//                        isInfoShown = true;
//                    }
//
//                    @Override
//                    public void delete() {
//                        BottomSheetUtils.showConfirmSheet(
//                                getSupportFragmentManager(),
//                                getString(R.string.Vault_DeleteFile_SheetTitle),
//                                getString(R.string.Vault_deleteFile_SheetDesc),
//                                getString(R.string.action_delete),
//                                getString(R.string.action_cancel),
//                                isConfirmed -> {
//                                    if (isConfirmed) {
//                                        viewerPresenter.confirmDeleteMediaFile(vaultFile);
//                                    }
//                                }
//                        );
//
//                    }
//                }
//        );
//    }
//
//    @Override
//    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
//        super.onActivityResult(requestCode, resultCode, data);
//        if (requestCode == PICKER_FILE_REQUEST_CODE) {
//            assert data != null;
//            viewerPresenter.exportNewMediaFile(withMetadata, handlingVaultFile, data.getData());
//        }
//    }
//
//    private void showShareWithMetadataDialog() {
//        LinkedHashMap<Integer, Integer> options = new LinkedHashMap<>();
//        options.put(1, R.string.verification_share_select_media_and_verification);
//        options.put(0, R.string.verification_share_select_only_media);
//
//        BottomSheetUtils.showRadioListOptionsSheet(
//                getSupportFragmentManager(),
//                getContext(),
//                options,
//                getString(R.string.verification_share_dialog_title),
//                getString(R.string.verification_share_dialog_expl),
//                getString(R.string.action_ok),
//                getString(R.string.action_cancel),
//                option -> startShareActivity(option > 0)
//        );
//    }
//
//    private void showExportWithMetadataDialog() {
//        LinkedHashMap<Integer, Integer> options = new LinkedHashMap<>();
//        options.put(1, R.string.verification_share_select_media_and_verification);
//        options.put(0, R.string.verification_share_select_only_media);
//        new Handler().post(() -> {
//            BottomSheetUtils.showRadioListOptionsSheet(
//                    getSupportFragmentManager(),
//                    getContext(),
//                    options,
//                    getString(R.string.verification_share_dialog_title),
//                    getString(R.string.verification_share_dialog_expl),
//                    getString(R.string.action_ok),
//                    getString(R.string.action_cancel),
//                    option -> {
//                        withMetadata = option > 0;
//                        maybeChangeTemporaryTimeout(() -> {
//                            performFileSearch();
//                            return Unit.INSTANCE;
//                        });
//                    }
//            );
//        });
//    }
//
//    private void initView() {
//        mPlay = binding.content.playAudio;
//        mRwd = binding.content.rwdButton;
//        mFwd = binding.content.rwdButton;
//        mTimer = binding.content.audioTime;
//        mDuration = binding.content.duration;
//        forward = binding.content.fwdButton;
//        rewind = binding.content.rwdButton;
//        toolbar = binding.toolbar;
//    }
//}

File: mobile/src/main/java/rs/readahead/washington/mobile/views/activity/viewer/AudioPlayActivity.kt
package rs.readahead.washington.mobile.views.activity.viewer

import android.app.ProgressDialog
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.view.WindowManager
import android.widget.ImageButton
import android.widget.TextView
import androidx.activity.viewModels
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.widget.Toolbar
import com.google.android.exoplayer2.ui.StyledPlayerView
import com.hzontal.tella_vault.Metadata
import com.hzontal.tella_vault.VaultFile
import org.hzontal.shared_ui.bottomsheet.BottomSheetUtils.ActionConfirmed
import org.hzontal.shared_ui.bottomsheet.BottomSheetUtils.showConfirmSheet
import org.hzontal.shared_ui.utils.DialogUtils
import rs.readahead.washington.mobile.MyApplication
import rs.readahead.washington.mobile.R
import rs.readahead.washington.mobile.bus.event.MediaFileDeletedEvent
import rs.readahead.washington.mobile.bus.event.VaultFileRenameEvent
import rs.readahead.washington.mobile.databinding.ActivityAudioPlayBinding
import rs.readahead.washington.mobile.media.AudioPlayer
import rs.readahead.washington.mobile.util.DialogsUtil
import rs.readahead.washington.mobile.util.ThreadUtil
import rs.readahead.washington.mobile.views.activity.MetadataViewerActivity
import rs.readahead.washington.mobile.views.activity.viewer.VaultActionsHelper.initContracts
import rs.readahead.washington.mobile.views.activity.viewer.VaultActionsHelper.showVaultActionsDialog
import rs.readahead.washington.mobile.views.base_ui.BaseLockActivity
import java.util.*
import java.util.concurrent.TimeUnit

class AudioPlayActivity : BaseLockActivity(), StyledPlayerView.ControllerVisibilityListener {
    var mPlay: ImageButton? = null
    var mRwd: ImageButton? = null
    var mFwd: ImageButton? = null
    var mTimer: TextView? = null
    var mDuration: TextView? = null
    var forward: View? = null
    var rewind: View? = null
    private var handlingVaultFile: VaultFile? = null
    private var audioPlayer: AudioPlayer? = null
    private var audioPlayerListener: AudioPlayer.Listener? = null

    private var showActions = false
    private var actionsDisabled = false
    private var withMetadata = false
    private var alertDialog: AlertDialog? = null
    private var progressDialog: ProgressDialog? = null
    private var paused = true
    private lateinit var toolbar: Toolbar
    private var isInfoShown = false
    private lateinit var binding: ActivityAudioPlayBinding
    private val viewModel: SharedMediaFileViewModel by viewModels()

    companion object {
        const val PLAY_MEDIA_FILE = "pmf"
        const val PLAY_MEDIA_FILE_ID_KEY = "pmfik"
        const val NO_ACTIONS = "na"
        private const val TIME_FORMAT = "%02d:%02d:%02d"
        private const val SEEK_DELAY = 15000
    }

    override fun onCreate(savedInstanceState: Bundle?) {

        super.onCreate(savedInstanceState)
        binding = ActivityAudioPlayBinding.inflate(layoutInflater)
        setContentView(binding.getRoot())
        actionsDisabled = intent.hasExtra(VideoViewerActivity.NO_ACTIONS)

        initView()
        initListeners()
        initContracts()
        setupToolbar()
        overridePendingTransition(R.anim.slide_in_start, R.anim.fade_out)
        enablePlay()
        initVaultMediaFile()
        initAudioListener()
        initObservers()
    }

    private fun initObservers() {
        with(viewModel) {
            error.observe(this@AudioPlayActivity) {
                onShowError(it)
            }
            onMediaFileExportStatus.observe(this@AudioPlayActivity) { status ->
                when (status) {
                    MediaFileExportStatus.EXPORT_START -> onExportStarted()
                    MediaFileExportStatus.EXPORT_PROGRESS -> onMediaExported()
                    MediaFileExportStatus.EXPORT_END -> onExportEnded()
                }
            }
            onMediaFileDeleted.observe(this@AudioPlayActivity) { deleted ->
                if (deleted) onMediaFileDeleted()
            }
            onMediaFileDeleteConfirmed.observe(this@AudioPlayActivity) { mediaFileDeletedConfirmation ->
                mediaFileDeletedConfirmation.vaultFile?.let { deletedVaultFile ->
                    onMediaFileDeleteConfirmation(
                        deletedVaultFile,
                        mediaFileDeletedConfirmation.showConfirmDelete
                    )
                }
            }
            onMediaFileRenamed.observe(this@AudioPlayActivity) { renamed ->
                onMediaFileRename(renamed)
            }
            onMediaFileGot.observe(this@AudioPlayActivity) { renamed ->
                onMediaFileSuccess(renamed)
            }
        }
    }

    private fun initVaultMediaFile() {
        if (intent.hasExtra(Companion.PLAY_MEDIA_FILE)) {
            val vaultFile =
                intent.getSerializableExtra(Companion.PLAY_MEDIA_FILE) as VaultFile?
            if (vaultFile != null) {
                ThreadUtil.runOnMain {
                    onMediaFileSuccess(
                        vaultFile
                    )
                }
            }
        } else if (intent.hasExtra(Companion.PLAY_MEDIA_FILE_ID_KEY)) {
            val id = intent.getStringExtra(Companion.PLAY_MEDIA_FILE_ID_KEY)
            if (id != null) {
                viewModel!!.getMediaFile(id)
            }
        }
    }

    private fun initAudioListener() {
        audioPlayerListener = object : AudioPlayer.Listener {
            override fun onStart(duration: Int) {
                mDuration!!.text = timeToString(duration.toLong())
            }

            override fun onStop() {
                stopPlayer()
                paused = true
                enablePlay()
                showTimeRemaining(0)
            }

            override fun onProgress(currentPosition: Int) {
                showTimeRemaining(currentPosition)
            }

            private fun showTimeRemaining(left: Int) {
                mTimer!!.text = timeToString(left.toLong())
            }
        }
    }

    override fun finish() {
        super.finish()
        overridePendingTransition(R.anim.slide_in_end, R.anim.slide_out_start)
    }


    private fun onShowError(errorResId: Int) {
        DialogUtils.showBottomMessage(
            this, getString(errorResId), true
        )
    }


    private fun initListeners() {
        mPlay!!.setOnClickListener {
            if (paused) {
                handlePlay()
            } else {
                handlePause()
            }
        }
        forward!!.setOnClickListener {
            audioPlayer!!.ffwd(
                SEEK_DELAY
            )
        }
        rewind!!.setOnClickListener {
            audioPlayer!!.rwd(
                SEEK_DELAY
            )
        }
    }

    override fun onBackPressed() {
        super.onBackPressed()
        if (isInfoShown) {
            toolbar.menu.findItem(R.id.menu_item_more).isVisible = true
            toolbar.menu.findItem(R.id.menu_item_metadata).isVisible = true
            toolbar.title = handlingVaultFile!!.name
        } else {
            stopPlayer()
            finish()
        }
    }

    override fun onResume() {
        super.onResume()
    }

    override fun onPause() {
        super.onPause()
        handlePause()
    }

    override fun onDestroy() {
        audioPlayerListener = null
        if (alertDialog != null && alertDialog!!.isShowing) {
            alertDialog!!.dismiss()
        }
        hideProgressDialog()
        super.onDestroy()
    }


    private fun onMediaFileSuccess(vaultFile: VaultFile) {
        handlingVaultFile = vaultFile
        toolbar.title = vaultFile.name
        //handlePlay();
        if (!actionsDisabled) {
            showActions = true
            invalidateOptionsMenu()
        }
    }

    fun onMediaExported() {
        showToast(resources.getQuantityString(R.plurals.gallery_toast_files_exported, 1, 1))
    }

    fun onExportError(error: Throwable) {
        showToast(R.string.gallery_toast_fail_exporting_to_device)
    }

    fun onExportStarted() {
        progressDialog = DialogsUtil.showProgressDialog(
            this,
            getString(R.string.gallery_save_to_device_dialog_progress_expl)
        )
    }

    fun onExportEnded() {
        hideProgressDialog()
    }

    private fun setupToolbar() {
        toolbar = binding.toolbar
        toolbar.setNavigationIcon(R.drawable.ic_arrow_back_white_24dp)
        toolbar.setNavigationOnClickListener {
            onBackPressed()
        }

        if (!actionsDisabled) {
            toolbar.inflateMenu(R.menu.video_view_menu)
            handlingVaultFile?.let { file ->
                setupMetadataMenuItem(file.metadata != null)
            }
            if (handlingVaultFile != null && handlingVaultFile!!.metadata != null) {
                val item = toolbar.menu.findItem(R.id.menu_item_metadata)
                item.isVisible = true
            }
            toolbar.menu.findItem(R.id.menu_item_more)
                .setOnMenuItemClickListener {
                    handlingVaultFile?.let { it1 ->
                        showVaultActionsDialog(
                            it1,
                            viewModel,
                            {
                                isInfoShown = true
                                onVisibilityChanged(View.VISIBLE)
                            },
                            toolbar = toolbar
                        )
                    }
                    false
                }
        }
    }

    private fun setupMetadataMenuItem(visible: Boolean) {
        if (actionsDisabled) {
            return
        }
        val mdMenuItem = toolbar.menu.findItem(R.id.menu_item_metadata)
        mdMenuItem.isVisible = visible
        if (visible) {
            mdMenuItem.setOnMenuItemClickListener {
                showMetadata()
                false
            }
        }
    }

    override fun onVisibilityChanged(visibility: Int) {
        toolbar.visibility = if (!isInfoShown) visibility else View.VISIBLE
    }


    private fun onMediaFileDeleteConfirmation(vaultFile: VaultFile, showConfirmDelete: Boolean) {
        if (showConfirmDelete) {
            showConfirmSheet(
                supportFragmentManager,
                getString(R.string.Vault_Warning_Title),
                getString(R.string.Vault_Confirm_delete_Description),
                getString(R.string.Vault_Delete_anyway),
                getString(R.string.action_cancel),
                object : ActionConfirmed {
                    override fun accept(isConfirmed: Boolean) {
                        if (isConfirmed) {
                            viewModel.deleteMediaFiles(vaultFile)
                        }
                    }
                }
            )
        } else {
            viewModel.deleteMediaFiles(vaultFile)
        }
    }

    fun onMediaFileDeleted() {
        MyApplication.bus().post(MediaFileDeletedEvent())
        finish()
    }

    fun onMediaFileDeletionError(throwable: Throwable) {
        showToast(R.string.gallery_toast_fail_deleting_files)
    }

    private fun onMediaFileRename(vaultFile: VaultFile) {
        toolbar.title = vaultFile.name
        MyApplication.bus().post(VaultFileRenameEvent())
    }

    fun onMediaFileRenameError(throwable: Throwable) {}

    fun getContext(): Context {
        return this
    }


    private fun handlePlay() {
        if (handlingVaultFile == null) {
            return
        }
        if (audioPlayer != null) {
            audioPlayer!!.resume()
        } else {
            audioPlayer = AudioPlayer(
                this,
                audioPlayerListener!!
            )
            audioPlayer!!.play(handlingVaultFile)
        }
        paused = false
        disablePlay()
        disableScreenTimeout()
    }

    private fun handlePause() {
        if (handlingVaultFile == null) {
            return
        }
        enablePlay()
        paused = true
        if (audioPlayer != null) {
            audioPlayer!!.pause()
        }
        enableScreenTimeout()
    }

    private fun onPlayerStop() {
        enablePlay()
    }

    private fun disablePlay() {
        mPlay!!.setImageDrawable(this.resources.getDrawable(R.drawable.big_white_pause_24p))
        enableButton(forward, mFwd)
        enableButton(rewind, mRwd)
    }

    private fun enablePlay() {
        mPlay!!.setImageDrawable(this.resources.getDrawable(R.drawable.ic_play_arrow_white_24dp))
        disableButton(forward, mFwd)
        disableButton(rewind, mRwd)
    }

    private fun enableButton(view: View?, button: ImageButton?) {
        button!!.isClickable = true
        view!!.alpha = 1f
    }

    private fun disableButton(view: View?, button: ImageButton?) {
        button!!.isClickable = false
        view!!.alpha = .3f
    }

    private fun stopPlayer() {
        if (audioPlayer != null) {
            audioPlayer!!.stop()
            audioPlayer = null
            onPlayerStop()
            enableScreenTimeout()
        }
    }

    private fun hideProgressDialog() {
        if (progressDialog != null) {
            progressDialog!!.dismiss()
            progressDialog = null
        }
    }

    private fun showMetadata() {
        val viewMetadata = Intent(this, MetadataViewerActivity::class.java)
        viewMetadata.putExtra(Metadata.VIEW_METADATA, handlingVaultFile)
        startActivity(viewMetadata)
    }

    private fun timeToString(duration: Long): String {
        return String.format(
            Locale.ROOT, Companion.TIME_FORMAT,
            TimeUnit.MILLISECONDS.toHours(duration),
            TimeUnit.MILLISECONDS.toMinutes(duration) -
                    TimeUnit.MINUTES.toMinutes(TimeUnit.MILLISECONDS.toHours(duration)),
            TimeUnit.MILLISECONDS.toSeconds(duration) -
                    TimeUnit.MINUTES.toSeconds(TimeUnit.MILLISECONDS.toMinutes(duration))
        )
    }

    private fun disableScreenTimeout() {
        window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
    }

    private fun enableScreenTimeout() {
        window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
    }

    private fun initView() {
        mPlay = binding!!.content.playAudio
        mRwd = binding!!.content.rwdButton
        mFwd = binding!!.content.rwdButton
        mTimer = binding!!.content.audioTime
        mDuration = binding!!.content.duration
        forward = binding!!.content.fwdButton
        rewind = binding!!.content.rwdButton
        toolbar = binding!!.toolbar
    }


}